File: 每日一题/216. 组合总和 III.cpp
class Solution
{
public:
    vector<vector<int>> ret;
    vector<int> temp;

    int dfs(int start, int end, int k, int n)
    {
        if (temp.size() > k || accumulate(temp.begin(), temp.end(), 0) > n)
        {
            return -1;
        }
        if (temp.size() == k && accumulate(temp.begin(), temp.end(), 0) == n)
        {
            ret.emplace_back(temp);
            return 1;
        }
        for (int x = start; x <= end; x++)
        {
            temp.push_back(x);
            dfs(x + 1, end, k, n);
            temp.pop_back();
        }
        return 0;
    }
    vector<vector<int>> combinationSum3(int k, int n)
    {
        dfs(1, 9, k, n);
        return ret;
    }
};

File: 每日一题/216. 组合总和 III.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        ret = []
        ans = []

        def dfs(st, en, n, k):
            if len(ans) > k or sum(ans) > n:
                return
            if len(ans) == k and sum(ans) == n:
                ret.append(ans[:])
                return
            for i in range(st, en+1):
                ans.append(i)
                dfs(i+1, en, n, k)
                ans.pop()
        dfs(1, 9, n, k)
        return ret
